### DISCUSSION TITLE
Callbacks onStart, onEnd, onProgress

### DISCUSSION THREAD

**Discussion by ahradnansky** [JAndrassy/ArduinoOTA#147] [Ideas]
Hi,
I must admit - great work! I use ArduinoOTA for all (possible) projects.
Would it be an extensive work to add above callbacks? onProgress is really not necessary..
OTA for ESP8266 has those and sometimes are useful.
Thx.

**Comment by JAndrassy** [comment]
there is 
```
  void beforeApply(void (*fn)(void)) 
  void onError(void (*fn)(int code, const char* msg)) 

```

**Comment by ahradnansky** [comment]
Yes, I know :-) In my case (that was ESP...) I had to disable some timer  interrupts to be able correcly finish upload.
Now, I have similar problem - ArduinoMega does quite a lot of comms via Ethernet (MQTT/PLC protocol gateway, telnet, etc.)
Upload starts, but never finishes (or finishes with error)...Idea is onStart to suppress all other task durint OTA update process.
In fact if I do OTA upload shortly after manual reset (when no all comms are established properly yet) it works...

**Comment by JAndrassy** [comment]
so add onStart, test it and then create a pull request
